src/TorchSharp/Tensor/Tensor.cs

```diff
@@ -881,6 +883,214 @@ public long[] size()
+            /// <remarks>The named tensor API is experimental and subject to change.</remarks>
+            public Tensor rename_(IEnumerable<string?>? names)
+            {
+                if (names != null && names.Count() != ndim && !names.Contains("..."))
+                    throw new ArgumentException($"The number of dimension names ({names.Count()}) is different from the number of dimensions ({ndim}).");
+
+                IntPtr res = IntPtr.Zero;
+
+                if (names != null && names.Count() > 0) {
+
+                    var dimNamesArray = ExpandEllipsis(names);
+
+                    IntPtr namesRef = IntPtr.Zero;
+
+                    using PinnedArray<IntPtr> pinnedArray = new PinnedArray<IntPtr>();
+                    namesRef = (names is null || names.Count() == 0) ? IntPtr.Zero : pinnedArray.CreateArray(dimNamesArray);
+
+                    res = THSTensor_rename_(Handle, namesRef, names is null ? 0 : dimNamesArray.Length);
+                } else {
+
+                    res = THSTensor_rename_(Handle, IntPtr.Zero, 0);
+                }
+
+                if (res == IntPtr.Zero) { torch.CheckForErrors(); }
+                return new Tensor(res);
+            }
+
+            /// <summary>
+            /// Refines the dimension names of the input tensor according to names.
+            /// 
+            /// Refining is a special case of renaming that “lifts” unnamed dimensions.A None dim can be refined to have any name; a named dim can only be refined to have the same name.
+            /// Because named tensors can coexist with unnamed tensors, refining names gives a nice way to write named-tensor-aware code that works with both named and unnamed tensors.
+            /// names may contain up to one ellipsis argument, passed as "...". The ellipsis is expanded greedily; it is expanded in-place to fill names to the same length as
+            /// this.shape using names from the corresponding indices of this.names.
+            /// </summary>
+            /// <param name="names">A list of names, one for each dimension in the tensor.</param>
+            /// <remarks>The named tensor API is experimental and subject to change.</remarks>
+            public Tensor refine_names(IEnumerable<string> names)
+            {
+                var dimNamesArray = ExpandEllipsis(names);
+
+                using PinnedArray<IntPtr> pinnedArray = new PinnedArray<IntPtr>();
+                IntPtr namesRef = pinnedArray.CreateArray(dimNamesArray);
+
+                IntPtr res = THSTensor_refine_names(Handle, namesRef, dimNamesArray.Length);
+                if (res == IntPtr.Zero) { torch.CheckForErrors(); }
+                return new Tensor(res);
+            }
+
+            private IntPtr[] ExpandEllipsis(IEnumerable<string?> names)
+            {
+                var namesArray = names.ToArray();
```

Comment:
>names.ToArray(); [](http://example.com/codeflow?start=33&length=16)

do we need this allocation? we can iterate over the IEnumerable. right? Also, IEnumerable support `Contains` method too. Or you think this is not hot path to optimize?

Reply:
No.

Reply:
The thing with this code is that there's a fair amount of indexing of several collections. It seemed easier to use indices everywhere, but I guess namesArray can remain an IEnumerable, actually. The code will be a little less readable, but not by much.